### PR TITLE
New filter that allows only study admins

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

Added a filter that only allows study admins to access the associate route.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [x] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java [hunk 1]
```diff
@@ -277,7 +275,7 @@ private static void start() {
 
         // before filter converts jwt into DDP_AUTH request attribute
         // we exclude the DSM paths. DSM paths have own separate authentication
-        before(API.BASE + "/*", new ExcludePathFilter(new TokenConverterFilter(new JWTConverter(userDao)),
+        before(API.BASE + "/*", new ExcludePathFilter(new TokenConverterFilter(new JWTConverter()),
                 API.DSM_BASE + "/*", API.CHECK_IRB_PASSWORD));
         before(API.BASE + "/*", new ExcludePathFilter(new AddDDPAuthLoggingFilter(),
                 API.DSM_BASE + "/*", API.CHECK_IRB_PASSWORD));
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java]: Was finding it easier to read the `beforeWithExclusion` method that already uses the ExcludePathFilter. Maybe we can get it back?

File: pepper-apis/src/test/resources/db-testscripts/baseline-seed-test-data.xml [hunk 2]
```diff
@@ -85,7 +85,7 @@
     </changeSet>
 
     <changeSet author="yufeng" id="baseline-seed-test-data-test-admin-user">
-        <validCheckSum>8:e473c4e24f73e2ffb441fd8dcc061de8</validCheckSum>
+        <validCheckSum>8:52721d61ce840801ee7b0000823e7dc2</validCheckSum>
```
Comment (MocanaAtBroad) [inline pepper-apis/src/test/resources/db-testscripts/baseline-seed-test-data.xml:88]: What happened here?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/security/UserPermissions.java [hunk 3]
```diff
@@ -129,6 +137,23 @@ public boolean canAccessUserProfile(String requestedUserGuid) {
...
+                Set<String> userStudies = handle.attach(JdbiUserStudyEnrollment.class)
+                        .getAllLatestEnrollmentsForUser(requestedUserGuid)
+                        .stream()
+                        .filter(status -> !status.getEnrollmentStatus().isExited())
+                        .map(EnrollmentStatusDto::getStudyGuid)
+                        .collect(Collectors.toSet());
+                for (var study : userStudies) {
+                    if (adminStudiesWithAccess.contains(study)) {
+                        canAccess = true;
+                        break;
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/security/UserPermissions.java]: `anyMatch`?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java [hunk 4]
```diff
@@ -565,6 +563,11 @@ public static void patch(String path, Route route, ResponseTransformer transform
         Spark.patch(path, route, transformer);
     }
 
+    public static void addBefore(Filter filter, String... paths) {
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java:588]: Don't think this is doing much for us right now.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java [hunk 5]
```diff
@@ -456,6 +452,8 @@ private static void start() {
         get(API.INVITATION, new GetInvitationRoute(), jsonSerializer);
         patch(API.INVITATION, new UpdateInvitationRoute(), jsonSerializer);
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java]: Should we be preventing exposing invitation codes on URLs?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/filter/OnlyStudyAdminFilter.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * A filter that only allows access to study admins.
+ */
+public class OnlyStudyAdminFilter implements Filter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(OnlyStudyAdminFilter.class);
+    private static final AuthPathRegexUtil pathMatcher = new AuthPathRegexUtil();
+
+    @Override
+    public void handle(Request request, Response response) {
+        DDPAuth ddpAuth = RouteUtil.getDDPAuth(request);
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/filter/OnlyStudyAdminFilter.java:24]: Have we established that our existing decoding of Auth0 tokens will work unchanged for BWH study admins? I believe were talking about authenticating CRCs against a different system. This is a question. I still don't know how this is supposed to work.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java (hunk 1 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java]: Added it back in the other PR.

File: pepper-apis/src/test/resources/db-testscripts/baseline-seed-test-data.xml (hunk 2 above)
Comment (yufengwng) [reply pepper-apis/src/test/resources/db-testscripts/baseline-seed-test-data.xml:88]: Removed the firecloud table inserts below, so the file checksum changed. Updated checksum here so liquibase doesn't complain.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java (hunk 4 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java:588]: Not right now, but will be adding more "study admin only" APIs, as outlined in design doc. Okay to keep it?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java (hunk 5 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java]: Update URLs in other PR.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/filter/OnlyStudyAdminFilter.java (hunk 6 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/filter/OnlyStudyAdminFilter.java:24]: Not sure. I will update as we learn more about that.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/security/UserPermissions.java (hunk 3 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/security/UserPermissions.java]: Good idea!

**Review by MocanaAtBroad — DISMISSED** [review]

**Review by MocanaAtBroad — DISMISSED** [review]

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**